noise_corrector.py

```diff
@@ -1,70 +1,177 @@
+  return [estimate(v) for v in observed_values]
+
+
+def generate_possible_outputs_per_source(max_reports, data_cardinality,
+                                         max_windows):
+  """Enumerates all possible outputs for any given source event
+
+  max_reports: The maximum number of triggers that can be attributed to this
+                source.
+  data_cardinality: The number of possible trigger_data values for the source.
+  max_windows: The number of reporting windows any given report can show up in.
+
+  Returns: a JSON list of all possible outputs. Each output is a list of size
+           `max_reports`, describing each report (its trigger_data and window).
+  """
+
+  def find_k_comb(N, k):
+    """Computes the Nth lexicographically smallest k-combination
+
+    https://en.wikipedia.org/wiki/Combinatorial_number_system
+
+    Follows the greedy algorithm outlined here:
+    http://math0.wvstateu.edu/~baker/cs405/code/Combinadics.html
+    """
+    target = N
+    c = 0
+    while math.comb(c, k) <= N:
+      c += 1
+
+    combos = []
+    while c >= 0 and k >= 0:
+      coef = math.comb(c, k)
+      if coef <= target:
+        combos.append(c)
+        k -= 1
+        target -= coef
+      c -= 1
+    return combos
+
+  def bars_to_report(star_index, num_star):
+    """Returns a report description"""
+
+    bars_prior_to_star = star_index - (max_reports - 1 - num_star)
+
+    # Any stars before any bars means the report is suppressed.
+    if bars_prior_to_star == 0:
+      return "No report"
+
+    # Otherwise, the number of bars just indexes into the cartesian product
+    # of |data_cardinality| x |num_windows|.
+    # Subtract the first bar because that just begins the first window.
+    window, data = divmod(bars_prior_to_star - 1, data_cardinality)
+    return {"window": window, "data": data}
+
+  def generate_output(N):
+    star_indices = find_k_comb(N, max_reports)
+    return [bars_to_report(s, i) for i, s in enumerate(star_indices)]
+
+  star_bar_length = data_cardinality * max_windows + max_reports
+  num_sequences = math.comb(star_bar_length, max_reports)
+  return [generate_output(i) for i in range(num_sequences)]
+
+
+def get_possible_event_source_outputs():
+  """Returns the possible outputs for an event source"""
+  return generate_possible_outputs_per_source(1, 2, 1)
+
+
+def get_possible_nav_source_outputs():
+  """Returns all the possible outputs for navigation source"""
+  return generate_possible_outputs_per_source(3, 8, 3)
+
+
+# Params are taken from the EVENT.md explainer:
+# https://github.com/WICG/conversion-measurement-api/blob/main/EVENT.md#data-limits-and-noise
+def correct_event_sources(observed_values):
+  # No triggers, 1 trigger with metadata 0, 1 trigger with metadata 1.
+  assert len(observed_values) == 3
+  randomized_response_rate = .0000025
+  return estimate_true_values(observed_values, randomized_response_rate)
+
+
+def correct_nav_sources(observed_values):
+  assert len(observed_values) == 2925
```

Comment:
Um it is fairly complicated to do this but I can try. PTAQL at what I have. I can give more details if it is useful.

Reply:
Thanks, I took a look. My understanding is that for event sources, this means the `noisy_reports` input array needs to have a length of 2925. 

Is this correct?

If not:
Can you help me understand the shape `noisy_reports` should have in both event and navigation case?

If yes:
Since devs will be providing and shaping the `noisy_reports` input, it means devs would need to understand the bars and stars to some extent, so that they provide a correct `noisy_reports` input to `correct_event_sources`.

In that case I think we'd want to:
* Explain bars and shapes a bit further, but probably in a separate write-up, the script may not be the right place for this.
OR
* Provide a utility function that takes as **input** actual reports⏤in the shape the reporting origin would receive them (JSON)⏤as provides as **output** report count arrays that have the right shape to be used as input to `correct_event_sources`. Would that be possible?
This could bridge the gap that IMU currently exists between the data the reporting endpoint would receive, and what `noisy_reports` (the array) actually should look like to be consumed properly by `correct_event_sources`.

(Note: I don't think this needs to be blocking for this script to be merged into this repo, that could be a follow-up)

Reply:
> Thanks, I took a look; my understanding is that for event sources, this means the input array to correct_nav_sources needs to have a length of 2925?
> 
> Is this correct?

I like your second option best, this seems the most useful from the developer's POV. It seems doable, I can look into it in a follow-up.

Reply:
Noted ty! 
I filed this an issue for reference: https://github.com/WICG/conversion-measurement-api/issues/301